### ISSUE TITLE
Support referencing converter instances for forcedTypes

### ISSUE THREAD

**Issue by runeflobakk** [jOOQ/jOOQ#13787]
### Use case:

I mainly use [qualified converters](https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/codegen-database-forced-types-converter/) for configuring forced types, as I prefer having the converter code in Java source files to have them checked by my IDE compiler. As the `<converter>` element refers to a class, this forces me to create a separate class for my converters by implementing `Converter`, or a bit more convenient, by extending `AbstractConverter`.

I have also noticed the `Converter.of(..)` and `Converter.ofNullable(..)` which offers a nice way to create converter _instances_ without all the separate class boilerplate. Though, as I have understood it, to configure these instances as a forcedType, one must use [inline converters](https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/codegen-database-forced-types-inline-converter/), i.e. write the Java code to create the instances in the configuration, which I am trying to avoid. A [lambda converter](https://www.jooq.org/doc/latest/manual/code-generation/codegen-advanced/codegen-config-database/codegen-database-forced-types/codegen-database-forced-types-lambda-converter/) is some helpful syntactic sugar, but it is still code, and it still feels a bit "clunky" to specify a lambda or method reference in the configuration.

Since I have a preference for using qualified converters, I have resorted to making my own "abstract converter" class, which enables me to define my converter classes in the equivalent manner as the `Converter.of(..)` factory methods, but by invoking a super constructor, like this:
```java
public static class MyIdConverter extends MyJooqAbstractConverter<UUID, MyId> {
    public MyIdConverter() {
        super(UUID.class, MyId::new, MyId.class, MyId::value);
    }
}
```
instead of overriding `from(T databaseObject)` and `to(U userObject)` of `AbstractConverter`.

### Possible solution you'd like to see:

This works, but what I think would offer me the best of both worlds of qualified converters and inline/lambda converters, would be something like this. In my Java code, I would like to be able to have the following:
```java
public record MyId(UUID value) {
    public static final Converter<UUID, MyId> JOOQ_CONVERTER = Converter.of(UUID.class, MyId.class, MyId::new, MyId::value);
}
```
And in the forced types configuration, I want to reference this _instance_ (instead of a class instantiated by jOOQ). Something like this:
```xml
<forcedType>
    <userType>my.app.MyId</userType>
    <converter>my.app.MyId.JOOQ_CONVERTER</converter>
    <includeExpression>my_table.id</includeExpression>
</forcedType>
```
The requirement for `my.app.MyId.JOOQ_CONVERTER` is to be a static field, would be nice if package private (default scope) is allowed. I would also be perfectly fine if a new element would be required for this, e.g. `<converterInstance>` or something, though I think perhaps it should be possible for the codegen tool to detect this?

Preferably, I would also like to be able to omit specifying `<userType>`, as this could be read from the field declaration.

A sidenote: I think the userType would be available also in the extending AbstractConverter case, as one is declaring the user type when declaring the concrete converter class, and this information is carried into the compiled code, and can actually be reflected on. There may of course be use cases I am not aware of for specifying other types (or perhaps more specific subtypes?) of the declared type in the converter class. Or the field declaration of my proposition above. But maybe it could work as a "fallback" to read if one chooses to _not_ configure a `<userType>` explicitly? This may be perhaps be an entirely separate issue.

### Possible workarounds:

I suppose the `MyJooqAbstractConverter<T, U>` abstract class (given in the first part of this issue description) can be viewed as a workaround :)

### Versions:

- jOOQ: 3.17.1 (and earlier)
- Java: Java 17 (and earlier)
- Database (include vendor): PostgreSQL
- OS: MacOS (any OS)
- JDBC Driver (include name if unofficial driver): postgresql 42.4.0

**Comment by lukaseder** [comment]
Thanks a lot for your feature request

> The requirement for `my.app.MyId.JOOQ_CONVERTER` is to be a static field, would be nice if package private (default scope) is allowed.

I think that the visibility isn't relevant to this feature request. If the converter is in the same package as the class(es) referencing it, then of course it can be package private. If it isn't in the same package (as per your own configuration), then it cannot be package private. There's not much jOOQ can do about this.

> I would also be perfectly fine if a new element would be required for this, e.g. `<converterInstance>` or something, though I think perhaps it should be possible for the codegen tool to detect this?

How would this be auto-detected? There's no way to tell from an identifier whether something is a class reference or a member reference. Sure, we could assume coding conventions, but what if we get it wrong? Someone will not like it, as we break their code.

> Preferably, I would also like to be able to omit specifying `<userType>`, as this could be read from the field declaration.

That would require compiling the `Converter` along with a correctly set up classpath, which is a can of worms that is not worth opening for a very minor convenience as the one suggested here.

Specifically, what is going to be solved here? Your `MyId` class is very generic, so it can apply to *every* ID column, I suspect? You're saving one line of configuration at the price of a very complex feature.

> A sidenote: I think the userType would be available also in the extending AbstractConverter case, as one is declaring the user type when declaring the concrete converter class, and this information is carried into the compiled code, and can actually be reflected on

"Reflected on" means that a compiled class is available, but we don't have that (in general). All of this is just String templates. It could even be kotlin/scala code, if you're using the `KotlinGenerator` or `ScalaGenerator`

This leaves us with the feature request of being able to supply a `<converterInstance>`, which is a static field reference, instead of `<converter>`, which is a class name reference. It could be done, yes, but I see this as a very low priority.

**Comment by runeflobakk** [comment]
I agree on all points. My confusion was that I did not think about which stage of the codegen happens, i.e. _before_ my own (and obviously the generated) code is actually compiled.
